toml/src/main/java/com/fasterxml/jackson/dataformat/toml/TomlGenerator.java

```diff
@@ -120,20 +120,23 @@ public Version version() {
 
     @Override
     public void close() throws IOException {
-        super.close();
-        _flushBuffer();
-        _outputTail = 0; // just to ensure we don't think there's anything buffered
-
-        if (_out != null) {
-            if (_ioContext.isResourceManaged() || isEnabled(StreamWriteFeature.AUTO_CLOSE_TARGET)) {
-                _out.close();
-            } else if (isEnabled(StreamWriteFeature.FLUSH_PASSED_TO_STREAM)) {
-                // If we can't close it, we should at least flush
-                _out.flush();
+        if (!isClosed()) {
+            super.close();
+            _flushBuffer();
+            _outputTail = 0; // just to ensure we don't think there's anything buffered
+
+            if (_out != null) {
+                if (_ioContext.isResourceManaged() || isEnabled(StreamWriteFeature.AUTO_CLOSE_TARGET)) {
+                    _out.close();
+                } else if (isEnabled(StreamWriteFeature.FLUSH_PASSED_TO_STREAM)) {
+                    // If we can't close it, we should at least flush
+                    _out.flush();
+                }
             }
+            _ioContext.close();
+            // Internal buffer(s) generator has can now be released as well
+            _releaseBuffers();
```

Comment:
This should be moved before `IOContext.close()` (not a huge deal, but it will likely return buffers into recycler, and that should happen before releasing recycler into pool).
I can move after merge.